### PR TITLE
Clean up provenance action section of provenance

### PR DESCRIPTION
**Pull request by MrCreosote**
* Add a builder
* Make immutable
* Add crosschecks between workspace object refs and their resolved counterparts

### CODE REVIEW

File: lib/Bio/KBase/workspace/Client.pm [hunk 1]
```diff
@@ -9556,8 +9556,8 @@ A provenance action.
         A provenance action (PA) is an action taken while transforming one data
```
Comment (MrCreosote) [inline lib/Bio/KBase/workspace/Client.pm:9556]: generated

File: lib/biokbase/workspace/client.py [hunk 2]
```diff
@@ -710,71 +710,72 @@ def save_objects(self, params, context=None):
            action. A provenance action (PA) is an action taken while
```
Comment (MrCreosote) [inline lib/biokbase/workspace/client.py:710]: generated

File: src/us/kbase/workspace/ProvenanceAction.java [hunk 3]
```diff
@@ -20,8 +20,8 @@
  *         A provenance action (PA) is an action taken while transforming one data
```
Comment (MrCreosote) [inline src/us/kbase/workspace/ProvenanceAction.java:20]: generated

File: src/us/kbase/workspace/database/Workspace.java [hunk 4]
```diff
@@ -836,20 +837,15 @@ private void sortObjects(
 			reports.put(wo, rep);
 			idhandler.associateObject(new IDAssociation(objcount, true));
 			try {
-				for (final Provenance.ProvenanceAction action: wo.getProvenance().getActions()) {
+				for (final ProvenanceAction action: wo.getProvenance().getActions()) {
 					for (final String pref: action.getWorkspaceObjects()) {
-						if (pref == null) {
```
Comment (MrCreosote) [inline src/us/kbase/workspace/database/Workspace.java:841]: provenance refs are now parsed earlier in the stack, so this error is impossible

File: src/us/kbase/workspace/kbase/ArgUtils.java [hunk 5]
```diff
@@ -194,9 +193,6 @@ public static Provenance processProvenance(final WorkspaceUser user,
 
 	// TODO CODE remove this eventually when everything uses Instants
 	private static String formatDate(final Date date) {
-		if (date == null) {
```
Comment (MrCreosote) [inline src/us/kbase/workspace/kbase/ArgUtils.java:197]: The remaining 2 methods that call this method never pass null

File: src/us/kbase/workspace/kbase/ArgUtils.java [hunk 6]
```diff
@@ -227,14 +223,7 @@ private static String formatDate(final Optional<Instant> date) {
 
 	private static List<UObject> translateMethodParametersToUObject(
 			final List<Object> methodParams) {
-		if (methodParams == null) {
```
Comment (MrCreosote) [inline src/us/kbase/workspace/kbase/ArgUtils.java:230]: methodParams are always returned as a list, not null, from ProvenanceAction

File: src/us/kbase/workspace/test/workspace/WorkspaceTest.java [hunk 7]
```diff
@@ -2602,31 +2601,11 @@ public void saveObjectWithTypeChecking() throws Exception {
...
-		failSave(userfoo, wspace, data, new TypedObjectValidationException(
-				"Object #2, obj4 has a null provenance reference"));
-		
-		badids = new Provenance(userfoo);
-		badids.addAction(new Provenance.ProvenanceAction().withWorkspaceObjects(Arrays.asList("")));
-		data.set(1, new WorkspaceSaveObject(obj4, data3, abstype0, null, badids, false));
-		failSave(userfoo, wspace, data, new TypedObjectValidationException(
-				"Object #2, obj4 has invalid provenance reference: IDs may not be null or the " +
-				"empty string"));
-
```
Comment (MrCreosote) [inline src/us/kbase/workspace/test/workspace/WorkspaceTest.java:2629]: These 3 tests are now pointless because bad refs are rejected by the provenance action class and are tested in the unit tests thereof. I added some integration tests in the JsonRPCLayerTest to exercise the error pathway though.

File: src/us/kbase/workspace/test/workspace/WorkspaceTest.java [hunk 8]
```diff
@@ -3568,23 +3554,23 @@ public void saveLargeProvenance() throws Exception {
...
 		
 		
 		methparams.add(TEXT1000);
 		Provenance p2 = new Provenance(foo);
-		p2.addAction(new ProvenanceAction().withMethodParameters(methparams));
+		p2.addAction(ProvenanceAction.getBuilder().withMethodParameters(methparams).build());
 		try {
 			ws.saveObjects(foo, prov, Arrays.asList(
-					new WorkspaceSaveObject(getRandomName(), data, SAFE_TYPE1, null, p, false)),
+					new WorkspaceSaveObject(getRandomName(), data, SAFE_TYPE1, null, p2, false)),
```
Comment (MrCreosote) [inline src/us/kbase/workspace/test/workspace/WorkspaceTest.java:3568]: This test has had a bug in it for years that should have caused it to fail - it was using the wrong provenance here. The reason the test passes when it shouldn't is because the provenance class is mutable.

File: src/us/kbase/workspace/test/workspace/WorkspaceTest.java [hunk 9]
```diff
@@ -3568,23 +3554,23 @@ public void saveLargeProvenance() throws Exception {
...
 					getIdFactory());
 			fail("saved too big prov");
 		} catch (IllegalArgumentException iae) {
 			assertThat("correct exception", iae.getLocalizedMessage(),
-					is(String.format("Object #1, %s provenance size 1000318 exceeds limit of " +
+					is(String.format("Object #1, %s provenance size 1000322 exceeds limit of " +
```
Comment (MrCreosote) [inline src/us/kbase/workspace/test/workspace/WorkspaceTest.java:3573]: Switched some `[]` to `null` in the json string that's used to calculate the size

File: src/us/kbase/workspace/test/workspace/WorkspaceTest.java [hunk 10]
```diff
@@ -3763,22 +3749,23 @@ public void saveWithReferencePaths() throws Exception {
...
+		Provenance prefs = new Provenance(u1).addAction(ProvenanceAction.getBuilder()
+				.withWorkspaceObjects(list(
 						"  \nreadws/4 ; 2/refref1/1;2/3 ;  	2/leaf1/1;  ",
-						"readws/readable")));
+						"readws/readable")).build());
 		saveObject(u1, testws, MT_MAP, MT_MAP, SAFE_TYPE1, "provtest1", prefs);
+		
 		checkProvenanceCorrect(
-				u1, prefs, testwsidb.withName("provtest1").build(), refPaths1);
+				u1, prefs, testwsidb.withName("provtest1").build(), trimKeys(refPaths1));
```
Comment (MrCreosote) [inline src/us/kbase/workspace/test/workspace/WorkspaceTest.java:3759]: Provenance refs are now `String.trim()`ed prior to entry to the workspace layer

File: src/us/kbase/workspace/test/workspace/WorkspaceTest.java [hunk 11]
```diff
@@ -3848,38 +3835,34 @@ public void saveWithReferencePaths() throws Exception {
...
 						"reference to 2/ref1/1"));
 		
-		// bad parse
-		pfail = new Provenance(u1).addAction(new ProvenanceAction()
-				.withWorkspaceObjects(Arrays.asList("readws/4;;2/refref1/1")));
-		failSave(u1, testws, fail, mt, type1, pfail,
-				new TypedObjectValidationException("Object #1, fail has unparseable provenance " +
-						"reference readws/4;;2/refref1/1: Reference path position 2: " +
-						"Illegal number of separators '/' in object reference ''"));
-
```
Comment (MrCreosote) [inline src/us/kbase/workspace/test/workspace/WorkspaceTest.java:3869]: Another now pointless test

File: src/us/kbase/workspace/database/Workspace.java [hunk 12]
```diff
@@ -836,20 +837,15 @@ private void sortObjects(
...
 					}
 				}
-			} catch (IdReferenceHandlerException ihre) {
-				throw new TypedObjectValidationException(String.format(
-						"Object %s has invalid provenance reference: ",
-						getObjectErrorId(wo, objcount)) + ihre.getMessage(), ihre);
+			} catch (IdReferenceHandlerException e) {
+				// thrown if ID is null or the empty string, which is impossible
+				throw new RuntimeException("This exception didn't actually happen. In fact "
+						+ "you're on extremely strong drugs. Pay me no heed", e);
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/Workspace.java:848]: I've always wondered about the best way to handle impossible errors. Now I know! 😁

Comment (MrCreosote) [reply src/us/kbase/workspace/database/Workspace.java:848]: An alternate approach: https://github.com/kbase/groups/blob/f9f17aacede77d71bf8d79ef98b5d653bc174163/src/us/kbase/groups/service/api/APICommon.java#L306-L307

File: src/us/kbase/workspace/test/kbase/JSONRPCLayerTester.java [hunk 13]
```diff
@@ -720,18 +720,13 @@ protected void checkProvenance(List<ProvenanceAction> expected,
...
-				while(gotmeth.hasNext()) {
-					assertThat("meth params equal", gotmeth.next().asClassInstance(Object.class),
-							is(expmeth.next().asClassInstance(Object.class)));
-				}
+			assertThat("method param count equal", gotpa.getMethodParams().size(),
+					is(exppa.getMethodParams().size()));
+			Iterator<UObject> gotmeth = gotpa.getMethodParams().iterator();
+			Iterator<UObject> expmeth = exppa.getMethodParams().iterator();
+			while (gotmeth.hasNext()) {
+				assertThat("meth params equal", gotmeth.next().asClassInstance(Object.class),
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/test/kbase/JSONRPCLayerTester.java:728]: "meth"... hur hur hur!

File: src/us/kbase/workspace/database/provenance/ProvenanceAction.java [hunk 14]
```diff
@@ -0,0 +1,744 @@
...
+		result = prime * result + ((service == null) ? 0 : service.hashCode());
+		result = prime * result + ((serviceVersion == null) ? 0 : serviceVersion.hashCode());
+		result = prime * result + ((subActions == null) ? 0 : subActions.hashCode());
+		result = prime * result + ((time == null) ? 0 : time.hashCode());
+		result = prime * result + ((wsobjs == null) ? 0 : wsobjs.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/provenance/ProvenanceAction.java:283]: This makes me itch for some sort of meta-object protocol that you could use to iterate through all the properties / attributes of the two objects, comparing them and exiting if a difference is found.

Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/ProvenanceAction.java:283]: The equals method is generated by Eclipse fwiw, so it's very low maintenance burden. I keep meaning to learn Lombok and have it generate hashCode and equals in bytecode but I never get around to it.

File: src/us/kbase/workspace/database/provenance/ProvenanceAction.java [hunk 15]
```diff
@@ -0,0 +1,744 @@
...
+					// if wsobjs is null, resolvedObjects must be null, so no need to test
+					incomingArgs == null &&
+					outgoingArgs == null &&
+					externalData == null &&
+					subActions == null &&
+					custom == null &&
+					description == null) {
+				// TODO PROV integration error test
+				throw new IllegalArgumentException(
+						"At least one field in a provenance action must be provided");
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/provenance/ProvenanceAction.java:736]: so demanding!

File: src/us/kbase/workspace/test/database/provenance/ProvenanceActionTest.java [hunk 16]
```diff
@@ -0,0 +1,1059 @@
...
+		} catch (Exception got) {
+			TestCommon.assertExceptionCorrect(got, new IllegalArgumentException(
+					"Null key in custom provenance"));
+		}
+	}
+	
+	@Test
+	public void buildFail() throws Exception {
+		try {
+			ProvenanceAction.getBuilder().build();
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/test/database/provenance/ProvenanceActionTest.java:1052]: Have you also tested the case where there are fields provided, but they're all null?

Comment (MrCreosote) [reply src/us/kbase/workspace/test/database/provenance/ProvenanceActionTest.java]: Yes. Yes I did

**Review by ialarmedalien — APPROVED** [review]
Looks as good as it gets!
